pubtools/pulplib/_impl/model/repository/repo_lock.py

```diff
@@ -0,0 +1,163 @@
+        activation_await_sleep=AWAIT_ACTIVATION_SLEEP,
+    ):
+        self._repo_id = repo_id
+        self._request_func = request_func
+        self._lock_claim = None
+        self._expiration_offset = expiration_offset
+        self._activation_await_sleep = activation_await_sleep
+
+    def __enter__(self):
+        LOG.info("Attempting to lock %s." % self._repo_id)
+        self.remove_expired_locks()
+        self._lock_claim = LockClaim(expiration_offset=self._expiration_offset)
+        self.submit_lock_claim()
+        self.await_lock_activation()
+
+    def __exit__(self, exc_type, exc_value, tb):
+        self.delete_lock_claim()
+        self._lock_claim = None
+
+    def get_repo_lock_data(self, filter_invalid=False):
+        LOG.debug("Getting current repo lock data for %s." % self._repo_id)
+        repo_url = REPO_ENDPOINT_BASE % self._repo_id
+
+        repo_raw_f = self._request_func(repo_url, method="GET")
+
+        notes_f = f_map(repo_raw_f, lambda data: (data.get("notes") or {}))
+        locks_f = f_map(
+            notes_f,
+            lambda notes: {key: notes[key] for key in notes if LOCK_CLAIM_STR in key},
+        )
+
+        locks_f = f_map(
+            locks_f,
+            lambda locks: [
+                LockClaim.from_json_data(json.loads(l)) for l in locks.values()
+            ],
+        )
+        parsed_locks = locks_f.result()
+        parsed_locks.sort()
+        if filter_invalid:
+            parsed_locks = [l for l in parsed_locks if l.is_valid]
+        return parsed_locks
+
+    def await_lock_activation(self):
+        sleep(max((self._lock_claim.valid_from - datetime.now()).seconds, 0))
+        while True:
+            lock_claims = self.get_repo_lock_data(filter_invalid=True)
+            # If our claim is the oldest valid lock claim.
+            if self._lock_claim == lock_claims[0]:
+                LOG.info("Lock obtained for %s." % self._repo_id)
+                return
+            else:
+                LOG.info("Awaiting lock for %s." % self._repo_id)
+                sleep(self._activation_await_sleep)
+
+    def submit_lock_claim(self):
+        LOG.info(
+            "Submitting lock with id '%s' to %s." % (self._lock_claim.id, self._repo_id)
+        )
+        self._send_note_update(
+            {LOCK_CLAIM_STR + self._lock_claim.id: self._lock_claim.as_json},
+        )
+
+    def delete_lock_claim(self):
+        LOG.info(
+            "Deleting lock with id '%s' from %s." % (self._lock_claim.id, self._repo_id)
+        )
+        self._send_note_update(
+            {LOCK_CLAIM_STR + self._lock_claim.id: None}, await_result=True
+        )
+
+    def remove_expired_locks(self):
+        invalid_locks_delta = {
+            LOCK_CLAIM_STR + lock.id: None
+            for lock in self.get_repo_lock_data()
+            if not lock.is_valid
+        }
+        if invalid_locks_delta:
+            LOG.debug("Removing invalid locks from %s." % self._repo_id)
```

Comment:
Given that this only happens if someone has crashed, I think this is worthy of WARNING log level, and probably mentioning every affected lock ID as well.